lib/quda_arpack_interface.cpp

```diff
@@ -1,384 +1,541 @@
+#ifdef ARPACK_LIB
 
-  struct SortEvals{
-    double _val;
-    int    _idx;
+  void arpackErrorHelpNAUPD();
+  void arpackErrorHelpNEUPD();
 
-    static bool small_values;
+#if (defined(QMP_COMMS) || defined(MPI_COMMS))
+#include <mpi.h>
+#endif
 
-    SortEvals(double val, int idx) : _val(val), _idx(idx) {};
+  void arpack_solve(void *h_evecs, void *h_evals, const DiracMatrix &mat, QudaEigParam *eig_param,
+                    ColorSpinorParam *cpuParam)
+  {
 
-    static bool SelectSmall (SortEvals v1, SortEvals v2) { return (v1._val < v2._val);}
-    static bool SelectLarge (SortEvals v1, SortEvals v2) { return (v1._val > v2._val);}
+    // ARPACK logfile name
+    char *arpack_logfile = eig_param->arpack_logfile;
+    if (getVerbosity() >= QUDA_SUMMARIZE) {
+      printfQuda("**** START ARPACK SOLUTION ****\n");
+      printfQuda("Output directed to %s\n", arpack_logfile);
+    }
 
-  };
+    // Create Eigensolver object for member function use
+    TimeProfile profile("Dummy");
+    EigenSolver *eig_solver = EigenSolver::create(eig_param, mat, profile);
+
+    // Construct parameters and memory allocation
+    //---------------------------------------------------------------------------------
+    double time_ar = 0.0; // time in ARPACK
+    double time_mv = 0.0; // time in QUDA mat vec + data transfer
+    double time_ev = 0.0; // time in computing Eigenvectors
+
+    // MPI objects
+    int *fcomm_ = nullptr;
+#if (defined(QMP_COMMS) || defined(MPI_COMMS))
+    MPI_Fint mpi_comm_fort = MPI_Comm_c2f(MPI_COMM_WORLD);
+    fcomm_ = static_cast<int *>(&mpi_comm_fort);
+#endif
 
-  bool SortEvals::small_values = true;
+    // Determine local volume for memory allocations
+    int local_dim[4];
+    int local_vol = 1;
+    for (int i = 0; i < 4; i++) {
+      local_dim[i] = cpuParam->x[i];
+      local_vol *= local_dim[i];
+    }
+    local_vol *= eig_param->invert_param->Ls;
+
+    int nSpin = (eig_param->invert_param->dslash_type == QUDA_LAPLACE_DSLASH) ? 1 : 4;
+
+    // all FORTRAN communication uses underscored
+    int ido_ = 0;
+    int info_ = 1; // if 0, use random vector. If 1, initial residual lives in resid_
+    int *ipntr_ = (int *)malloc(14 * sizeof(int));
+    int *iparam_ = (int *)malloc(11 * sizeof(int));
+    int n_ = local_vol * nSpin * 3, nEv_ = eig_param->nEv, nKr_ = eig_param->nKr, ldv_ = local_vol * nSpin * 3,
+        lworkl_ = (3 * nKr_ * nKr_ + 5 * nKr_) * 2, rvec_ = 1;
+    int max_iter = eig_param->max_restarts * (nKr_ - nEv_) + nEv_;
+    int *h_evals_sorted_idx = (int *)malloc(nKr_ * sizeof(int));
+
+    // Assign values to ARPACK params
+    iparam_[0] = 1;
+    iparam_[2] = max_iter;
+    iparam_[3] = 1;
+    iparam_[6] = 1;
+
+    // ARPACK problem type to be solved
+    char howmny = 'P';
+    char bmat = 'I';
+    char *spectrum;
+    spectrum = strdup("SR"); // Initialsed just to stop the compiler warning...
+
+    if (eig_param->use_poly_acc) {
+      if (eig_param->spectrum == QUDA_SPECTRUM_SR_EIG)
```

Comment:
See comment I've put elsewhere about throwing warnings/errors about polynomial acceleration for non-normal ops

Reply:
Again, leaving this alone. For normal operators checks, see here: https://github.com/lattice/quda/blob/feature/deflatedMG/lib/interface_quda.cpp#L2470

Reply:
Marking this as resolved...